### DISCUSSION TITLE
Feature Request: Make UI more customizable

### DISCUSSION THREAD

**Discussion by 413x1nkp** [zed-industries/zed#24904] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

I think others would agree that having a more customizable UI is something that a lot of people want.
As a user, I'd love to be able to customize my IDE, not by just adding NEW stuff, but by modifying what existing stuff I would, and wouldn't like to keep.
Case and point, I think a lot of people can agree with me on the fact that not even half of Zed users are using it solely because it has chatgpt/LLM/copilot built-in. In fact, I'd argue that it's just pure bloat that shouldn't be included by default. However, this feature request is more about general customization and modification of the IDE users can perform, like hiding specific buttons, displaying more/less info on the dock, replacing hamburger menu with something else, etc.

The way I think would be the best to implement that is doing it the way VSCode/VSCodium does it, by just having options on what to display where. Of course you could go and do it in fancier way, for example: like Librewolf, which allows users to build/modify the UI on-the-spot by clicking a single button and then just drag-n-dropping stuff the user wants to keep or discard.

### If applicable, add mockups / screenshots to help present your vision of the feature

Let's say, I'd like to change this default UI layout
![1](https://github.com/user-attachments/assets/3226a69b-e2f4-42fe-8c47-8cb4411e3ae2)

Maybe I, the user, would like to change the UI to not have a seemingly redundant panel below the file tabs/buffers, like so
![2](https://github.com/user-attachments/assets/6941513d-a5ab-4e60-94b7-e3278090d226)

Not all of the users would like these buttons either, since those perform the same action as some hotkeys, which means that those who use keyboard for navigation will still be forced to stare at these
![image](https://github.com/user-attachments/assets/e33d0f0a-99cc-4915-b96e-b5af7fb9719f)

So, let's say a user would also want to get rid of those, here's my proposition on the look
![3](https://github.com/user-attachments/assets/ee80dbd3-4429-43bf-9132-235f4a5686bd)

Now, displaying the name of the file opened TWICE seems a bit excessive for my taste, especially when only a SINGLE file is opened. Let's say, for cases when just ONE file is opened, it will only be highlighted in the tree-sitter. This will make the UI look more clean, in my opinion
![4](https://github.com/user-attachments/assets/12b52c17-c271-4792-8733-4b27e7401aec)

Let's take a look at these buttons in the now empty panel which is supposed to display opened tabs/buffers.
All of these (except the last one) have hotkeys that can be used instead of manually clicking on them with a mouse, which is much more appealing to me personally; therefore, I believe there should also be options on configuring those as well, as I don't see any scenario in which I'll use these buttons instead of regular hotkeys on a keyboard.
![image](https://github.com/user-attachments/assets/c021e95c-777f-48ea-bc2b-5c945b9363f2)

The last button in particular seems to be completely useless for someone like me, and the hotkey under which it is bound genuinely confuses me
![5](https://github.com/user-attachments/assets/c467e3d0-2c0b-4f5d-b71f-163d2d4efa63)

Now, the last bunch are probably the less significant ones, but that doesn't mean they should be ignored. For instance, the hamburger menu contains only the commands which can also be executed by simply using command palette (CTRL+SHIFT+P OR F1), therefore, having an option to simply hide it would also be very nice.
![image](https://github.com/user-attachments/assets/5abdb09b-84bd-4fa4-a3ec-9278cc692540)

Now we can notice another problem with repetitiveness! The project's name is repeated twice! I understand the idea behind showing opened projects as a folder within the tree-sitter, however if You are working on a single project at a time, with only the single project open, this just seems extremely redundant. So ONLY FOR CASES WHEN JUST A SINGLE PROJECT IS OPENED, displaying project as a folder in the tree-sitter doesn't make too much sense.
I believe changing the way projects are shown in the tree-sitter to be like this will look more clean for people that work on one project at a time
![image](https://github.com/user-attachments/assets/b9d3bfce-7e32-4e2d-b1bf-b78a9fce912e)

Now, if we can get rid of the three buttons (the ones on the tabs dock), we can consider working on the open tabs/buffers dock. We could theoretically embed it within the top most dock, which would save a lot of space
![7](https://github.com/user-attachments/assets/af8c5f46-c53d-4b32-adc9-8ef4a506057b)

We could also add an extra touch of centering the project name, just as a flex
![image](https://github.com/user-attachments/assets/9110b1c4-5cb6-48e7-9c1d-1349fcde7fce)

Now, all of this can only be possible if UI customization becomes a thing, which I hope eventually does. I personally would prefer working in the mocked up version of the Zed made by me more than I'd prefer working in the stock UI version. Just as a finishing part, here's the before and after the "customization" of Zed I'd personally perform if I could:
- Before
![1](https://github.com/user-attachments/assets/3226a69b-e2f4-42fe-8c47-8cb4411e3ae2)

- After
![8](https://github.com/user-attachments/assets/4ed0222f-281d-4001-b434-06cb19580ca9)

# Conclusion
While in reality I **barely** touched on customization possibilities that I and others would enjoy to have, I believe it is a valid thing to want in an IDE, even if the customization ends up just being able to hide the stock buttons or re-arrange them, which still would be a huge plus, since as I explained earlier, some of the stuff just seems redundant and some features are used by just a handful of others, such as the LLM/copilot thing, which, mind you, constantly takes some portion of your IDE (no matter how small it is), despite You possibly not even planning on ever using it.

**Comment by notpeter** [comment]
There's a lot here.  Filing as a single issue makes it difficult to parse, but as I read it you're requesting (at least) the following:

- [x] Ability to hide breadcrumbs bar (Possible today: see breadcrumbs setting below).
- [ ] Hide forward/back toolbar icons
- [ ] Hide tab bar completely (when single file opened)
- [ ] Hide "new" toolbar button
- [ ] Hide "split pane" toolbar button
- [ ] Hide "zoom in" toolbar button (the strange icon "⎋" is for "esc" (matching MacOS iconography); shortcut is shift+esc)
- [ ] Hide workspace folder from Project panel when only one folder in workspace
- [ ] Merge tab bar into title bar

We currently support the following in your settings.json:
```
  "toolbar": {
    // Whether to show breadcrumbs.
    "breadcrumbs": false,
    // Whether to show quick action buttons.
    "quick_actions": false,
    // Whether to show the Selections menu in the editor toolbar
    "selections_menu": false
  },
```

I think some of your ideas could be implemented as additional settings options for hiding toolbar icons, while others are less likely to be implemented (hiding the hamburger, centering the workspace indictor/selector, etc).

You may be interested in:
- #4382

P.S. When referring to the "project panel" you incorrectly used the phrase "tree-sitter" (understandable, they are both "trees") but that's a [whole different thing](https://tree-sitter.github.io/tree-sitter/).

**Comment by bbb651** [comment]
Hiding forward/back already exists as well, you can set `tab_bar.show_nav_history_buttons` to `false`.
